src/hardware/rohde-schwarz-sme-0x/protocol.c

```diff
@@ -24,66 +25,368 @@
+	ret = sr_scpi_get_double(sdi->conn, cmd, power);
+	if (ret != SR_OK)
 		return SR_ERR;
 
 	return SR_OK;
 }
 
+SR_PRIV int rs_sme0x_get_clk_src_idx(const struct sr_dev_inst *sdi, int *idx)
+{
+	char *buf;
+	int ret;
+	struct dev_context *devc;
+	const char *cmd;
+	const char *resp_int;
+	const char *resp_ext;
+
+	if (!sdi || !sdi->priv)
+		return SR_ERR;
+	devc = sdi->priv;
+
+	cmd = devc->model_config->commands[RS_CMD_GET_CLK_SRC];
+	ret = sr_scpi_get_string(sdi->conn, cmd, &buf);
+	if (ret != SR_OK || !buf)
+		return SR_ERR;
+
+	resp_int = devc->model_config->responses[RS_RESP_CLK_SRC_INT];
+	resp_ext = devc->model_config->responses[RS_RESP_CLK_SRC_EXT];
+	if (strcmp(buf, resp_int) == 0) {
+		ret = SR_OK;
+		*idx = 0;
+	} else if (strcmp(buf, resp_ext) == 0) {
+		ret = SR_OK;
+		*idx = 1;
+	} else {
+		ret = SR_ERR;
+	}
+
+	g_free(buf);
+	return ret;
+}
+
+SR_PRIV int rs_sme0x_set_enable(const struct sr_dev_inst *sdi, gboolean enable)
+{
+	struct dev_context *devc;
+
+	if (!sdi || !sdi->priv)
+		return SR_ERR;
+	devc = sdi->priv;
+
+	if (devc->enable == enable)
+		return SR_OK;
+
+	devc->enable = enable;
+
+	return sr_scpi_send(sdi->conn, (devc->enable) ?
+					devc->model_config->commands[RS_CMD_SET_ENABLE] :
+					devc->model_config->commands[RS_CMD_SET_DISABLE]);
+}
+
 SR_PRIV int rs_sme0x_set_freq(const struct sr_dev_inst *sdi, double freq)
 {
 	struct dev_context *devc;
-	const struct rs_device_model *config;
 
+	if (!sdi || !sdi->priv)
+		return SR_ERR;
 	devc = sdi->priv;
-	config = devc->model_config;
 
-	if ((freq > config->freq_max) || (freq < config->freq_min))
+	if ((freq > devc->freq_max) || (freq < devc->freq_min))
 		return SR_ERR_ARG;
 
-	return sr_scpi_send(sdi->conn, commands[RS_CMD_SET_FREQ], freq);
+	return sr_scpi_send(sdi->conn,
+						devc->model_config->commands[RS_CMD_SET_FREQ], freq);
 }
 
 SR_PRIV int rs_sme0x_set_power(const struct sr_dev_inst *sdi, double power)
 {
 	struct dev_context *devc;
-	const struct rs_device_model *config;
 
+	if (!sdi || !sdi->priv)
+		return SR_ERR;
 	devc = sdi->priv;
-	config = devc->model_config;
 
-	if ((power > config->power_max) || (power < config->power_min))
+	if ((power > devc->power_max) || (power < devc->power_min))
 		return SR_ERR_ARG;
 
-	return sr_scpi_send(sdi->conn, commands[RS_CMD_SET_POWER], power);
+	return sr_scpi_send(sdi->conn,
+						devc->model_config->commands[RS_CMD_SET_POWER], power);
+}
+
+SR_PRIV int rs_sme0x_set_clk_src(const struct sr_dev_inst *sdi, int idx)
+{
+	struct dev_context *devc;
+
+	if (!sdi || !sdi->priv || !sdi->conn)
+		return SR_ERR;
+	devc = sdi->priv;
+
+	if (devc->clk_source_idx == idx)
+		return SR_OK;
+
+	devc->clk_source_idx = idx;
+
+	return sr_scpi_send(sdi->conn, (idx == 0) ?
+					devc->model_config->commands[RS_CMD_SET_CLK_SRC_INT] :
+					devc->model_config->commands[RS_CMD_SET_CLK_SRC_EXT]);
+}
+
+SR_PRIV int rs_sme0x_get_minmax_freq(const struct sr_dev_inst *sdi,
+									double *min_freq, double *max_freq)
```

Comment:
That's hard to read, just indent it with one tab (same for `rs_sme0x_get_minmax_power`).